### PR TITLE
UCT/IB/MLX5: Use proper MMIO mode for DEVX

### LINKED ISSUE #5001: shm_ib/test_ucp_rma_mt.put_get/1 failure on legacy test node
```
2020-04-08T16:12:07.6348612Z [----------] 1 test from shm_ib/test_ucp_rma_mt
2020-04-08T16:12:07.6349779Z [ RUN      ] shm_ib/test_ucp_rma_mt.put_get/1 <shm,ib>
2020-04-08T16:12:24.1050625Z /scrap/azure/agent-09/AZP_WORKSPACE/2/s/contrib/../test/gtest/ucp/test_ucp_rma_mt.cc:175: Failure
2020-04-08T16:12:24.1158005Z Value of: target_data[i]
2020-04-08T16:12:24.1159851Z   Actual: 16045690984833335043
2020-04-08T16:12:24.1160627Z Expected: orig_data[i]
2020-04-08T16:12:24.1161313Z Which is: 0
2020-04-08T16:12:24.1162965Z /scrap/azure/agent-09/AZP_WORKSPACE/2/s/contrib/../test/gtest/ucp/test_ucp_rma_mt.cc:175: Failure
2020-04-08T16:12:24.1163966Z Value of: target_data[i]
2020-04-08T16:12:24.1164695Z   Actual: 16045690984833335023
2020-04-08T16:12:24.1165436Z Expected: orig_data[i]
2020-04-08T16:12:24.1166116Z Which is: 0
2020-04-08T16:12:24.1167543Z /scrap/azure/agent-09/AZP_WORKSPACE/2/s/contrib/../test/gtest/ucp/test_ucp_rma_mt.cc:175: Failure
2020-04-08T16:12:24.1168549Z Value of: target_data[i]
2020-04-08T16:12:24.1169276Z   Actual: 16045690984833335033
2020-04-08T16:12:24.1170017Z Expected: orig_data[i]
2020-04-08T16:12:25.7996278Z Which is: 0
2020-04-08T16:12:25.7997756Z /scrap/azure/agent-09/AZP_WORKSPACE/2/s/contrib/../test/gtest/ucp/test_ucp_rma_mt.cc:175: Failure
2020-04-08T16:12:25.7998675Z Value of: target_data[i]
2020-04-08T16:12:25.7999078Z   Actual: 16045690984833335053
2020-04-08T16:12:25.7999461Z Expected: orig_data[i]
2020-04-08T16:12:25.7999801Z Which is: 0
2020-04-08T16:12:25.8000658Z /scrap/azure/agent-09/AZP_WORKSPACE/2/s/contrib/../test/gtest/ucp/test_ucp_rma_mt.cc:197: Failure
2020-04-08T16:12:25.8001276Z Error: Endpoint timeout
2020-04-08T16:12:25.8002134Z /scrap/azure/agent-09/AZP_WORKSPACE/2/s/contrib/../test/gtest/ucp/test_ucp_rma_mt.cc:197: Failure
2020-04-08T16:12:25.8002726Z Error: Endpoint timeout
2020-04-08T16:12:25.8003516Z terminate called after throwing an instance of 'ucs::test_abort_exception'
2020-04-08T16:12:25.8004075Z   what():  std::exception
2020-04-08T16:12:25.8004944Z [swx-rdmz-ucx-legacy-01:57501:1:3010] Caught signal 6 (Aborted: tkill(2) or tgkill(2))
2020-04-08T16:12:25.8005567Z terminate called recursively
2020-04-08T16:12:25.8005938Z make: *** [test] Aborted
2020-04-08T16:12:25.8006825Z make: Leaving directory `/scrap/azure/agent-09/AZP_WORKSPACE/2/s/build-test/test/gtest'
```
https://dev.azure.com/ucfconsort/ucx/_build/results?buildId=5326&view=logs&j=88d29b50-4e11-53d5-399b-e634502852e6&t=cb773578-1ba5-52b8-cb10-437b22e73aab
https://dev.azure.com/ucfconsort/0b36e3f0-8ab9-4a48-b68b-4b2350e02c88/_apis/build/builds/5326/logs/95

### PR DESCRIPTION
**Pull request by Artemy-Mellanox**
Fixes #5001

### CODE REVIEW

File: src/uct/ib/ud/accel/ud_mlx5.c [hunk 1]
```diff
@@ -716,7 +716,7 @@ static ucs_status_t uct_ud_mlx5_iface_create_qp(uct_ib_iface_t *ib_iface,
     uct_ib_mlx5_qp_t *qp = &iface->tx.wq.super;
     ucs_status_t status;
 
-    status = uct_ib_mlx5_iface_create_qp(ib_iface, qp, attr);
+    status = uct_ib_mlx5_iface_create_qp(ib_iface, qp, (uct_ib_mlx5_qp_attr_t *)attr);
```
Comment (yosefe) [inline src/uct/ib/ud/accel/ud_mlx5.c:719]: why need to cast? can we pass uct_ib_mlx5_qp_attr_t * to this function?

File: src/uct/ib/mlx5/dv/ib_mlx5_dv.c [hunk 2]
```diff
@@ -50,21 +50,28 @@ ucs_status_t uct_ib_mlx5_devx_create_qp(uct_ib_iface_t *iface,
...
     int max_tx, max_rx, len_tx, len;
     int wqe_size;
     int dvflags;
     void *qpc;
     int ret;
 
-    uct_ib_iface_fill_attr(iface, attr);
+    uct_ib_iface_fill_attr(iface, &attr->super);
+
+    status = uct_ib_mlx5_get_mmio_mode(iface->super.worker, attr->mmio_mode,
```
Comment (yosefe) [inline src/uct/ib/mlx5/dv/ib_mlx5_dv.c:63]: can we make setting mmio mode part of uct_rc_mlx5_iface_fill_attr()?

Comment (Artemy-Mellanox) [reply src/uct/ib/mlx5/dv/ib_mlx5_dv.c:63]: not really
fill_attr goes before qp create
while in verbs mmio mode set after qp create

File: src/uct/ib/ud/accel/ud_mlx5.c (hunk 1 above)
Comment (Artemy-Mellanox) [reply src/uct/ib/ud/accel/ud_mlx5.c:719]: because ud_iface_ops.create_qp is callback common to verbs and mlx5

Comment (yosefe) [reply src/uct/ib/ud/accel/ud_mlx5.c:719]: can we use ucs_derived_of ?

Comment (Artemy-Mellanox) [reply src/uct/ib/ud/accel/ud_mlx5.c:719]: We can, but I would prefer allocate new uct_ib_mlx5_qp_attr_t in this function and initialize it with upper one.
```
diff --git a/src/uct/ib/ud/accel/ud_mlx5.c b/src/uct/ib/ud/accel/ud_mlx5.c
index 2528fef..6769f3d 100644
--- a/src/uct/ib/ud/accel/ud_mlx5.c
+++ b/src/uct/ib/ud/accel/ud_mlx5.c
@@ -709,14 +709,17 @@ static void uct_ud_mlx5_iface_event_cq(uct_ib_iface_t *ib_iface,
 }

 static ucs_status_t uct_ud_mlx5_iface_create_qp(uct_ib_iface_t *ib_iface,
-                                                uct_ib_qp_attr_t *attr,
+                                                uct_ib_qp_attr_t *ib_attr,
                                                 struct ibv_qp **qp_p)
 {
     uct_ud_mlx5_iface_t *iface = ucs_derived_of(ib_iface, uct_ud_mlx5_iface_t);
     uct_ib_mlx5_qp_t *qp = &iface->tx.wq.super;
+    uct_ib_mlx5_qp_attr_t attr = {};
     ucs_status_t status;

-    status = uct_ib_mlx5_iface_create_qp(ib_iface, qp, (uct_ib_mlx5_qp_attr_t *)attr);
+    attr.super = *ib_attr;
+
+    status = uct_ib_mlx5_iface_create_qp(ib_iface, qp, &attr);
     if (status != UCS_OK) {
         return status;
     }
```

Comment (yosefe) [reply src/uct/ib/ud/accel/ud_mlx5.c:719]: in this suggestion, where is the mlx5 part (mmio mode) initialized?

Comment (Artemy-Mellanox) [reply src/uct/ib/ud/accel/ud_mlx5.c:719]: it's not
UD has no DEVX flow
mmio set [later](https://github.com/openucx/ucx/blob/master/src/uct/ib/ud/accel/ud_mlx5.c#L818) in iface constructor

Comment (yosefe) [reply src/uct/ib/ud/accel/ud_mlx5.c:719]: better not pass uninitialized struct to a function, it's recipe for unexpected bugs
if a function does not need a parameter - it should not be passed at all

Comment (Artemy-Mellanox) [reply src/uct/ib/ud/accel/ud_mlx5.c:719]: my point was to pass it initialized to zero

Comment (yosefe) [reply src/uct/ib/ud/accel/ud_mlx5.c:719]: but what is the meaning of mmio_mode=0? is it really what we want as default?

Comment (Artemy-Mellanox) [reply src/uct/ib/ud/accel/ud_mlx5.c:719]: indeed. let's set to UCT_IB_MLX5_MMIO_MODE_LAST

Comment (yosefe) [reply src/uct/ib/ud/accel/ud_mlx5.c:719]: and assert that mode!= LAST when it's used?

File: src/uct/ib/mlx5/ib_mlx5.c [hunk 3]
```diff
@@ -277,7 +277,7 @@ void uct_ib_mlx5_iface_put_res_domain(uct_ib_mlx5_qp_t *qp)
 
 ucs_status_t uct_ib_mlx5_iface_create_qp(uct_ib_iface_t *iface,
                                          uct_ib_mlx5_qp_t *qp,
-                                         uct_ib_qp_attr_t *attr)
+                                         uct_ib_mlx5_qp_attr_t *attr)
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5.c:280]: is this function using mmio_mode? or maybe it can accept just .super part?

**Comment by yosefe** [comment]
issue is #4975
bot:pipe:retest

**Pull request merged**